hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/shortcircuit/ShortCircuitCache.java

```diff
@@ -181,25 +182,49 @@ public long getRateInMs() {
 
     @Override
     public void run() {
+      if (slot == null) {
+        return;
+      }
       LOG.trace("{}: about to release {}", ShortCircuitCache.this, slot);
       final DfsClientShm shm = (DfsClientShm)slot.getShm();
       final DomainSocket shmSock = shm.getPeer().getDomainSocket();
       final String path = shmSock.getPath();
+      DataOutputStream out = null;
       boolean success = false;
-      try (DomainSocket sock = DomainSocket.connect(path);
-           DataOutputStream out = new DataOutputStream(
-               new BufferedOutputStream(sock.getOutputStream()))) {
-        new Sender(out).releaseShortCircuitFds(slot.getSlotId());
-        DataInputStream in = new DataInputStream(sock.getInputStream());
-        ReleaseShortCircuitAccessResponseProto resp =
-            ReleaseShortCircuitAccessResponseProto.parseFrom(
-                PBHelperClient.vintPrefixed(in));
-        if (resp.getStatus() != Status.SUCCESS) {
-          String error = resp.hasError() ? resp.getError() : "(unknown)";
-          throw new IOException(resp.getStatus().toString() + ": " + error);
+      int retries = 2;
+      try {
+        while (retries > 0) {
+          try {
+            if (domainSocket == null || !domainSocket.isOpen()) {
+              // we are running in single thread mode, no protection needed for
```

Comment:
not that familiar with short circuit read. Is true that it's single threaded?

Reply:
yeah, the SlotReleaser is running in single thread as follow code.
 ShortCircuitCache#uref->replica.close()->cache.scheduleSlotReleaser(slot)->releaserExecutor.execute(new SlotReleaser(slot))-> SlotReleaser#run {
...
try {
            if (domainSocket == null || !domainSocket.isOpen()) {
              // we are running in single thread mode, no protection needed for
              // domainSocket
              domainSocket = DomainSocket.connect(path);
            }
...
}
Since ShortCircuitCache#uref run in lock，this code is run serially in a core thread of a thread pool.

Reply:
yeah
ShortCircuitCache#unref->ShortCircuitReplica#close->cache.scheduleSlotReleaser(slot)->releaserExecutor.execute(new SlotReleaser(slot))->SlotReleaser#run {
...
if (domainSocket == null || !domainSocket.isOpen()) {
              // we are running in single thread mode, no protection needed for
              // domainSocket
              domainSocket = DomainSocket.connect(path);
            }
...
}

Since ShortCircuitCache#unref run in lock, this code is run serially in a thread of a thread pool.

Reply:
Got it. The release scheduler has only one thread.
/**
   * The executor service that runs the cacheCleaner.
   */
  private final ScheduledThreadPoolExecutor releaserExecutor
      = new ScheduledThreadPoolExecutor(1, new ThreadFactoryBuilder().
      setDaemon(true).setNameFormat("ShortCircuitCache_SlotReleaser").
      build());